Cache block asset resource hints

joshuatf opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR caches the block asset resource hints.  The `get_script_dependency_src_array` calls are recursive and result in about 10ms load on the page when items exist in the cart.  Caching the resource hints (and dependencies) lessens this load.

**Before:** 78.249ms
**After:** 67.691ms

This gives us a reduction of 36% over baseline when cart items are present when compared to a baseline local site without WooCommerce as measured in https://github.com/woocommerce/woocommerce/issues/50714.

Closes #49230  .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

#### Regression testing

1. Clear website cache.
2. Go to the homepage with nothing in your cart.
3. View source. Look for rel='prefetch' rules. Cart and checkout will not be there.
4. Go to the shop page and add something to the cart.
5. Go back to the homepage.
6. View source. Look for rel='prefetch' rules:

![149147708-009cf5fa-46fa-4cbb-80c6-6fe7f0cd5607](https://github.com/user-attachments/assets/70d32489-56fa-4d9c-8147-15e7697b7027)

#### Performance testing

1. Add the following snippet to your site to ensure a product always exists in the cart, replacing `PRODUCT_ID` with a real product id from your site:
```php
add_action( 'wp_loaded', function() {
	if ( wc()->cart && wc()->cart->is_empty() ) {
		wc()->cart->add_to_cart( PRODUCT_ID );
	}
} );
```
2. In a terminal, check the TTFB: `curl -o /dev/null -w "Connect: %{time_connect} TTFB: %{time_starttransfer} Total time: %{time_total} \n" testsite.local`, replacing `testsite.local` with your site.
3. Switch to `trunk` and run the above command again.
4. Note the drop in TTFB on this branch compared to `trunk`
5. This command can be repeated multiple times or you can use a script like [TTFB](https://github.com/jaygooby/ttfb.sh) to generate a larger sample size

#### Development testing

1. Modify some dependencies for the cached resource hints (e.g., remove the array of dependencies in `assets/client/blocks/cart-frontend.asset.php`)
2. In your `wp-config.php` set the development mode to `plugin`.  `define( 'WP_DEVELOPMENT_MODE', 'plugin' );`
3. Also add `define( 'SCRIPT_DEBUG', true );` to your `wp-config.php` (this is for cache busting the script dependency API)
4. Add something to your cart
5. Check the source for the page
7. Make sure that the cache is not used and the cart `prefetch` resources are no longer listed

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

joshuatf committed:
Cache block asset resource hints

joshuatf committed:
Combine resource hints under single cache transient

joshuatf committed:
Add AssetController tests

joshuatf committed:
Add changelog entry

mikejolley commented on plugins/woocommerce/src/Blocks/AssetsController.php:
```diff
@@ -196,13 +196,59 @@ private function get_prerender_resource_hints() {
+			return $cached[ $filename ];
+		}
+
 		if ( ! $filename ) {
```
This seems like it should be before the cache lookup since that requires a filename.

mikejolley commented on plugins/woocommerce/src/Blocks/AssetsController.php:
```diff
@@ -196,13 +196,59 @@ private function get_prerender_resource_hints() {
+			'wordpress'   => get_bloginfo( 'version' ),
+		);
+
+		if ( isset( $cache['version'] ) && $cache['version'] === $current_version ) {
```
Could this cause confusion in development builds if dependencies change during dev?

joshuatf replied:
Good call.  Added a dev check in https://github.com/woocommerce/woocommerce/pull/51874/commits/a8a59dc339338424dafc94e1649ef53fcee1000e.

Note that this uses the newer `WP_DEVELOPMENT_MODE` which I believe is slightly more appropriate than `SCRIPT_DEBUG` for caching, but may create some minor confusion as we transition between these modes.  I'll add the following testing instructions above as well:

#### Development testing

1. Modify some dependencies for the cached resource hints (e.g., remove the array of dependencies in `assets/client/blocks/cart-frontend.asset.php`)
2. In your `wp-config.php` set the development mode to `plugin`.  `define( 'WP_DEVELOPMENT_MODE', 'plugin' );`
3. Also add `define( 'SCRIPT_DEBUG', true );` to your `wp-config.php` (this is for cache busting the script dependency API)
4. Add something to your cart
5. Check the source for the page
6. Make sure that the cache is not used and the cart `prefetch` resources are no longer listed

mikejolley reviewed (changes requested):
Tests well. I added some questions and there are linting errors concerning indentation.

Thanks

joshuatf committed:
Skip cache use in development mode

joshuatf committed:
Fix up lint errors

joshuatf commented:
Thanks for the review, @mikejolley!  Added some testing instructions around the development cache busting and fixed up the linting errors as well.  This should be ready for another review.

mikejolley commented on plugins/woocommerce/src/Blocks/AssetsController.php:
```diff
@@ -31,7 +33,7 @@ public function __construct( AssetApi $asset_api ) {
 	 * Initialize class features.
 	 */
-	protected function init() {
+	public function init() {
```
I am aware of a new linting rule `WooCommerce.Functions.InternalInjectionMethod.MissingPublic` which is why this was changed, but I am not 100% sure it applies to this use case. There are other places in codebase where the rule was just ignored. Once this goes public it cannot go back so I'd recommend leaving protected.

joshuatf replied:
Good call.  Reverted to protected.

mikejolley commented on plugins/woocommerce/src/Blocks/AssetsController.php:
```diff
@@ -31,7 +33,7 @@ public function __construct( AssetApi $asset_api ) {
 	 * Initialize class features.
 	 */
-	protected function init() {
+	public function init() {
```
```suggestion
	protected function init() { // phpcs:ignore WooCommerce.Functions.InternalInjectionMethod.MissingPublic
```

mikejolley commented on plugins/woocommerce/src/Blocks/AssetsController.php:
```diff
@@ -196,6 +198,50 @@ private function get_prerender_resource_hints() {
+	 * @return array|null Array of resource hints.
+	 */
+	private function get_block_asset_resource_hints_cache() {
+		if ( wp_is_development_mode( 'plugin' ) ) {
```
Not seen this before 👍🏻

joshuatf replied:
It was my first time using this as well.  [Relatively new and introduced in 6.3](https://make.wordpress.org/core/2023/07/14/configuring-development-mode-in-6-3/).

It avoids some of the ambiguity around `SCRIPT_DEBUG` and `WP_DEBUG` that we previously used in development.

mikejolley reviewed (commented):
@joshuatf In the development mode testing instructions it reads:

> Make sure that the cache is not used and the cart prefetch resources are no longer listed

I still get prefetch and I think thats correct. Can you confirm what I should be seeing?

joshuatf committed:
Revert init method back to protected and ignore linting

Co-authored-by: Mike Jolley <mike.jolley@me.com>

joshuatf commented:
>I still get prefetch and I think thats correct. Can you confirm what I should be seeing?

@mikejolley This line is in follow up to the previous suggestion to remove the dependencies:

>Modify some dependencies for the cached resource hints (e.g., remove the array of dependencies in assets/client/blocks/cart-frontend.asset.php)

If those dependencies are removed, they should no longer be prefetched and the cache should not be used if we are in development mode.

Happy to update this testing instruction if it's unclear.

mikejolley reviewed (approved):
Thanks for those changes. Approving.

Regarding the test case, prefetching didn't change with the cart script. I edited `checkout-frontend.assets.php`, then with an item in cart I checked the prefetched scripts on the homepage. Hope this was correct.

joshuatf commented:
> Regarding the test case, prefetching didn't change with the cart script. I edited checkout-frontend.assets.php, then with an item in cart I checked the prefetched scripts on the homepage. Hope this was correct.

Discussed this with @mikejolley on Slack and this appears to be working as expected.  There's a need to remove dependencies from `cart-frontend.assets.php` and `checkout-frontend.assets.php` to test development changes if the removed dependency exists across both scripts (since one script would keep the dependency prefetched).  Removing or adding a script that is not overlapped should be immediately reflected in the source code.
